### PR TITLE
Motorola Surnia Backlight Driver

### PR DESCRIPTION
**Pull request by BogdanTheGeek**
Added support for brightness control for motorola-surnia (probably works for osprey and styx too) though led-backlight driver and rohm bd65b60 driver.

### CODE REVIEW

**Comment by BogdanTheGeek** [comment]
Pretty sure include/linux/platform_data/leds-bd65b60.h can me moved into the source file.

File: Documentation/devicetree/bindings/leds/rohm,bd65b60.txt [hunk 1]
```diff
@@ -0,0 +1,47 @@
+ROHM BD65B60 device tree bindings
```
Comment (TravMurav) [inline Documentation/devicetree/bindings/leds/rohm,bd65b60.txt:1]: plaintext bindings are deprecated and the new style yaml bindings must be used for newly added drivers

File: Documentation/devicetree/bindings/leds/rohm,bd65b60.txt [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+rohm,init-level   : Default brightness [0, 255].
+rohm,ovp-val   : Over Voltage Protection Detect Voltage (can be: 0x0, 0x8 or 0x10)
+
+Example
+=======
+
+	wled: bd65b60@64 {
+		compatible = "rohm,bd65b60";
+		reg = <0x64>;
+		linux,name = "wled:backlight";
```
Comment (TravMurav) [inline Documentation/devicetree/bindings/leds/rohm,bd65b60.txt:40]: I believe now it's color+function and this one is deprecated

File: Documentation/devicetree/bindings/leds/rohm,bd65b60.txt [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+rohm,default-on   : Set LED to on state by default.
+rohm,led1-used : LED1 is being used.
+rohm,led2-used : LED2 is being used.
+rohm,init-level   : Default brightness [0, 255].
+rohm,ovp-val   : Over Voltage Protection Detect Voltage (can be: 0x0, 0x8 or 0x10)
+
+Example
+=======
+
+	wled: bd65b60@64 {
```
Comment (TravMurav) [inline Documentation/devicetree/bindings/leds/rohm,bd65b60.txt:37]:
```suggestion
	wled: led-controller@64 {
```
node names must be generic per DT spec

File: arch/arm64/boot/dts/qcom/msm8916-motorola-surnia.dts [hunk 4]
```diff
@@ -57,6 +57,28 @@
 		pinctrl-1 = <&usb_id_sleep>;
 		pinctrl-names = "default", "sleep";
 	};
+
+	backlight: panel-backlight {
```
Comment (TravMurav) [inline arch/arm64/boot/dts/qcom/msm8916-motorola-surnia.dts:61]:
```suggestion
	backlight: backlight {
```

File: arch/arm64/boot/dts/qcom/msm8916-motorola-surnia.dts [hunk 5]
```diff
@@ -57,6 +57,28 @@
...
+		compatible = "led-backlight";
+		leds = <&wled>;
+		default-brightness-level = <192>;
+	   };
+};
+
+&blsp_i2c1 {
+	status = "okay";
+
+	wled: bd65b60@64 {
```
Comment (TravMurav) [inline arch/arm64/boot/dts/qcom/msm8916-motorola-surnia.dts:71]: ditto

File: arch/arm64/boot/dts/qcom/msm8916-motorola-surnia.dts [hunk 6]
```diff
@@ -341,6 +371,13 @@
 		bias-pull-down;
 	};
 
+	ts_reset_default: ts-reset-default-state {
```
Comment (TravMurav) [inline arch/arm64/boot/dts/qcom/msm8916-motorola-surnia.dts:374]: what uses this?

File: drivers/leds/leds-bd65b60.c [hunk 7]
```diff
@@ -0,0 +1,318 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
```
Comment (TravMurav) [inline drivers/leds/leds-bd65b60.c:1]: driver must be a separate commit from the dt. Three commits:
- Add the bindigns doc
- Add the driver
- Enable the backlight for the device (add a user for the driver)

File: drivers/leds/leds-bd65b60.c [hunk 8]
```diff
@@ -0,0 +1,318 @@
...
+#include <linux/module.h>
+#include <linux/slab.h>
+#include <linux/i2c.h>
+#include <linux/leds.h>
+#include <linux/err.h>
+#include <linux/delay.h>
+#include <linux/uaccess.h>
+#include <linux/interrupt.h>
+#include <linux/regmap.h>
+#include <linux/platform_data/leds-bd65b60.h>
```
Comment (TravMurav) [inline drivers/leds/leds-bd65b60.c:18]: let's merge these into a single file

File: drivers/leds/leds-bd65b60.c [hunk 9]
```diff
@@ -0,0 +1,318 @@
...
+static void bd65b60_brightness_set(struct work_struct *work)
+{
+	struct bd65b60_chip *pchip =
+		container_of(work, struct bd65b60_chip, ledwork);
+	unsigned int level = pchip->cdev.brightness;
+	static int old_level = -1;
+
+	/* set configure pwm input on first brightness command */
+	if (old_level == -1) {
+		dev_info(pchip->dev, "Enabling CABC");
```
Comment (TravMurav) [inline drivers/leds/leds-bd65b60.c:121]: Avoid spamming the dmesg, there is no reason to create a message every time you turn the screen on

File: drivers/leds/leds-bd65b60.c [hunk 10]
```diff
@@ -0,0 +1,318 @@
...
+}
+
+#ifdef CONFIG_OF
+static int bd65b60_dt_init(struct i2c_client *client,
+			   struct bd65b60_platform_data *pdata)
+{
+	struct device_node *np = client->dev.of_node;
+	int rc;
+
+	rc = of_property_read_string(np, "linux,name", &pdata->name);
```
Comment (TravMurav) [inline drivers/leds/leds-bd65b60.c:175]: I believe the core should pick the node name and this should be deprecated.

File: drivers/leds/leds-bd65b60.c [hunk 11]
```diff
@@ -0,0 +1,318 @@
...
+	}
+
+	if (value > led_cdev->max_brightness)
+		value = led_cdev->max_brightness;
+
+	led_cdev->brightness = value;
+	schedule_work(&pchip->ledwork);
+}
+
+#ifdef CONFIG_OF
```
Comment (TravMurav) [inline drivers/leds/leds-bd65b60.c:168]: I believe you don't have to guard for the OF and the core has the nops for the of methods

File: drivers/leds/leds-bd65b60.c [hunk 12]
```diff
@@ -0,0 +1,318 @@
...
+		pdata->led_sel |= BD65B60_LED1SEL;
+	if (of_property_read_bool(np, "rohm,led2-used"))
+		pdata->led_sel |= BD65B60_LED2SEL;
+
+	pdata->no_reset = of_property_read_bool(np, "rohm,no-reset");
+
+	pdata->default_on = of_property_read_bool(np, "rohm,default-on");
+	pdata->init_level = BD65B60_DEFAULT_BRIGHTNESS;
+	if (pdata->default_on)
+		of_property_read_u32(np, "rohm,init-level", &pdata->init_level);
```
Comment (TravMurav) [inline drivers/leds/leds-bd65b60.c:194]: this sounds like default-state = (on|off|keep) from the generic binding (see Documentation/devicetree/bindings/leds/common.yaml)

File: drivers/leds/leds-bd65b60.c [hunk 13]
```diff
@@ -0,0 +1,318 @@
...
+	pdata->ovp_val = BD65B60_DEFAULT_OVP_VAL;
+	of_property_read_u32(np, "rohm,ovp-val", &pdata->ovp_val);
+
+	return 0;
+}
+
+static const struct of_device_id of_bd65b60_leds_match[] = {
+	{
+		.compatible = "rohm,bd65b60",
+	},
```
Comment (TravMurav) [inline drivers/leds/leds-bd65b60.c:205]:
```suggestion
	{ .compatible = "rohm,bd65b60", },
```

File: drivers/leds/leds-bd65b60.c [hunk 14]
```diff
@@ -0,0 +1,318 @@
...
+
+	return 0;
+}
+
+static void bd65b60_remove(struct i2c_client *client)
+{
+	struct bd65b60_chip *pchip = i2c_get_clientdata(client);
+
+	bd65b60_write(pchip, REG_PON, 0);
+	led_classdev_unregister(&pchip->cdev);
```
Comment (TravMurav) [inline drivers/leds/leds-bd65b60.c:296]: you use dev-managed register so the core should unregister it on it's own when the device is removed

File: drivers/leds/leds-bd65b60.c [hunk 15]
```diff
@@ -0,0 +1,318 @@
...
+}
+
+static const struct i2c_device_id bd65b60_id[] = { { BD65B60_NAME, 0 }, {} };
```
Comment (TravMurav) [inline drivers/leds/leds-bd65b60.c:299]:
```suggestion
static const struct i2c_device_id bd65b60_id[] = {
	{ BD65B60_NAME, 0 },
	{},
};
```

and maybe just put the string from the define here if nothing else uses it

File: drivers/leds/leds-bd65b60.c [hunk 16]
```diff
@@ -0,0 +1,318 @@
...
+
+MODULE_DEVICE_TABLE(i2c, bd65b60_id);
+MODULE_DEVICE_TABLE(of, of_bd65b60_leds_match);
```
Comment (TravMurav) [inline drivers/leds/leds-bd65b60.c:302]: Please move these right after respective arrays,

File: arch/arm64/boot/dts/qcom/msm8916-motorola-surnia.dts [hunk 17]
```diff
@@ -320,6 +342,14 @@
 		bias-pull-up;
 	};
 
+	backlight_reset_default: backlight-reset-default-state {
+		pins = "gpio36";
+		function = "gpio";
+		drive-strength = <2>;
+		bias-pull-down;
+		output-high;
```
Comment (TravMurav) [inline arch/arm64/boot/dts/qcom/msm8916-motorola-surnia.dts:350]: tbh I'd prefer the driver to assert this gpio

File: drivers/leds/Makefile [hunk 18]
```diff
@@ -86,6 +86,7 @@ obj-$(CONFIG_LEDS_TURRIS_OMNIA)		+= leds-turris-omnia.o
 obj-$(CONFIG_LEDS_WM831X_STATUS)	+= leds-wm831x-status.o
 obj-$(CONFIG_LEDS_WM8350)		+= leds-wm8350.o
 obj-$(CONFIG_LEDS_WRAP)			+= leds-wrap.o
+obj-$(CONFIG_LEDS_BD65B60)		+= leds-bd65b60.o
```
Comment (TravMurav) [inline drivers/leds/Makefile:89]: probably want to sort alphabetically

File: drivers/leds/leds-bd65b60.c [hunk 19]
```diff
@@ -0,0 +1,318 @@
...
+
+static const struct of_device_id of_bd65b60_leds_match;
+
+#endif
+
+static int bd65b60_probe(struct i2c_client *client,
+			 const struct i2c_device_id *id)
+{
+	struct device_node *np = client->dev.of_node;
+	struct bd65b60_platform_data *pdata = dev_get_platdata(&client->dev);
```
Comment (TravMurav) [inline drivers/leds/leds-bd65b60.c:224]: platform_data sounds like board files, let's drop support for those fossils

**Comment by BogdanTheGeek** [comment]
looks like i basically have to re-write this driver to make it fall in line with the led-controller standard. I've got a bit to learn, so ill work on it on a different branch and rebase when its ready for review if thats ok.

**Comment by BogdanTheGeek** [comment]
Resubmitted in #296
